### DISCUSSION TITLE
XVNC Switching Layouts

### DISCUSSION THREAD

**Discussion by mr-fv** [neutrinolabs/xrdp#3094] [Q&A] [answered]
**The keyboard layout does not switch in XVNC.**

I use two layouts us and ru. Added the ability to switch layouts using this [wiki](https://github.com/neutrinolabs/xrdp/wiki/Example-of-xrdp_keyboard.ini-for-russian-keyboard). All other settings are by default. The second layout appeared on the panel. But when switching to the ru language, characters from the layout that was connected to the xrdp server (us), are still printed.

I exported the us and ru layouts:
`xkbcomp $DISPLAY keyboard.xkb`
_— both files are correct._
[us.xkb.txt](https://github.com/neutrinolabs/xrdp/files/15468132/us.xkb.txt)
[ru.xkb.txt](https://github.com/neutrinolabs/xrdp/files/15468130/ru.xkb.txt)

I looked at the output of the currently pressed keys via the xev command in the terminal.

_On the us layout:_
`
KeyRelease event, serial 37, synthetic NO, window 0x4eO0001, root 0x2a5, subw 0x4e000O2, time 5691621, (52,29), root:(793,470), state 0x10, keycode 24 (keysym 0x71, q), same_screen YES, 
XLookupString gives 1 bytes: (71) "q" 
XFilterEvent returns: False
`

_On the ru layout:_
`
KeyRelease event, serial 43, synthetic NO, window 0x5200001, root 0x2a5, subw 0x5200002, time 4608536, (38,26), root: (811,461), state 0x2010, keycode 253 (keysym 0x71, q), same_screen YES, 
XKeysymToKeycode returns keycode: 24 
XLookupString gives 1 bytes: (71) "q" 
XFilterEvent returns: False
`

In xev, everything is correct on the us layout. The ru layout shows that the keycode number 253 is defined, which corresponds to a void in xkb. And with further pressing of different keys, empty xkb_symbols will be filled in us letters. But at the same time, a line appears in xev XKeysymToKeycode, which shows really the keycodes that was pressed. And according to ru.xkb, this code should have been equal to keysym `<AD01> Cyrillic_shorti`.

**Is it possible to implement two layouts in xvnc xrdp?**

_Cinnamon session
xrdp-0.9.21-1.el7_
[sesman.ini.txt](https://github.com/neutrinolabs/xrdp/files/15468153/sesman.ini.txt)
[startwm.sh.txt](https://github.com/neutrinolabs/xrdp/files/15468156/startwm.sh.txt)

**Comment by matt335672** [comment, marked as answer]
It may not be at the moment.

The VNC server receives a symbol from us rather than a key code, which is what you need.

More information from the TigerVNC server side can be found here:-

https://github.com/TigerVNC/tigervnc/issues/962

The link mentions a TigerVNC option called `-RawKeyboard` which could solve this, but we don't support it. The protocol reference for that is here:-

https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#qemu-extended-key-event-message
